File: packages/flow-dev-tools/src/check-test/checkTestConfig.js
/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 * @format
 */

const fs = require('fs');
const {join} = require('path');

export type TestConfig = {
  auto_start: boolean,
  shell: string,
  cmd: string,
  stdin: string,
  ignore_stderr: boolean,
  cwd: string,
  file_watcher: string,
  start_args: string,
  wait_for_recheck: string,
  skip_saved_state: boolean,
  saved_state_only: boolean,
  git: boolean,
  skip_rust_port: boolean,
};

async function parseTestConfig(testDir: string): Promise<TestConfig> {
  const config: TestConfig = {
    auto_start: true,
    shell: '',
    cmd: 'full-check',
    stdin: '',
    ignore_stderr: true,
    cwd: '',
    file_watcher: 'none',
    start_args: '',
    wait_for_recheck: 'true',
    skip_saved_state: false,
    saved_state_only: false,
    git: false,
    skip_rust_port: false,
  };

  const configPath = join(testDir, '.testconfig');
  try {
    await fs.promises.access(configPath);
  } catch {
    return config;
  }

  const content = await fs.promises.readFile(configPath, 'utf8');
  const lines = content.split('\n');

  // Track whether cmd was explicitly set in the config file.
  // In bash, shell is processed first (clears cmd), then cmd is processed
  // (overrides). This means cmd always wins regardless of file order.
  // We match that by deferring the shell→cmd clearing to after the loop.
  let cmdExplicitlySet = false;

  for (const line of lines) {
    const trimmed = line.trim();
    if (trimmed === '' || trimmed.startsWith('#')) continue;

    const colonIndex = trimmed.indexOf(':');
    if (colonIndex === -1) continue;

    const key = trimmed.substring(0, colonIndex).trim();
    const value = trimmed.substring(colonIndex + 1).trim();

    switch (key) {
      case 'auto_start':
        config.auto_start = value !== 'false';
        break;
      case 'shell':
        config.shell = value;
        break;
      case 'cmd': {
        // cmd takes the rest of the line after "cmd:" with whitespace trimmed.
        // Use .trim() (not .replace(/^ /, '')) to handle any amount of
        // whitespace after the colon, matching bash's awk field-splitting
        // which normalizes leading whitespace.
        config.cmd = trimmed.substring(colonIndex + 1).trim();
        cmdExplicitlySet = true;
        break;
      }
      case 'stdin':
        config.stdin = value;
        break;
      case 'ignore_stderr':
        config.ignore_stderr = value !== 'false';
        break;
      case 'cwd':
        config.cwd = value;
        break;
      case 'file_watcher':
        if (value !== '') {
          config.file_watcher = value;
        }
        break;
      case 'start_args': {
        // start_args takes everything after "start_args:" with whitespace
        // trimmed. Use .trim() to match bash's awk field-splitting which
        // normalizes leading whitespace. Without this, extra spaces after
        // the colon leave residual whitespace that causes split(/\s+/) to
        // produce empty array elements.
        config.start_args = trimmed.substring(colonIndex + 1).trim();
        break;
      }
      case 'wait_for_recheck':
        config.wait_for_recheck = value === 'false' ? 'false' : 'true';
        break;
      case 'skip_saved_state':
        config.skip_saved_state = value === 'true';
        break;
      case 'saved_state_only':
        config.saved_state_only = value === 'true';
        break;
      case 'git':
        config.git = value === 'true';
        break;
      case 'skip_rust_port':
        config.skip_rust_port = value === 'true';
        break;
    }
  }

  // Match bash semantics: if shell is set but cmd was NOT explicitly
  // overridden, clear cmd. If cmd was explicitly set, it wins.
  if (config.shell !== '' && !cmdExplicitlySet) {
    config.cmd = '';
  }

  return config;
}

module.exports = {parseTestConfig};

File: rust_port/crates/flow_dev_tools/src/runtests/check_test_config.rs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use std::fs;
use std::io;
use std::path::Path;

use super::exists;

#[derive(Clone, Debug, Eq, PartialEq)]
pub(super) struct TestConfig {
    pub(super) auto_start: bool,
    pub(super) shell: String,
    pub(super) cmd: String,
    pub(super) stdin: String,
    pub(super) ignore_stderr: bool,
    pub(super) cwd: String,
    pub(super) file_watcher: String,
    pub(super) start_args: String,
    pub(super) wait_for_recheck: String,
    pub(super) skip_saved_state: bool,
    pub(super) saved_state_only: bool,
    pub(super) git: bool,
    pub(super) skip_windows: bool,
}

pub(super) fn parse_test_config(test_dir: &Path) -> io::Result<TestConfig> {
    let mut config = TestConfig {
        auto_start: true,
        shell: String::new(),
        cmd: "full-check".to_owned(),
        stdin: String::new(),
        ignore_stderr: true,
        cwd: String::new(),
        file_watcher: "none".to_owned(),
        start_args: String::new(),
        wait_for_recheck: "true".to_owned(),
        skip_saved_state: false,
        saved_state_only: false,
        git: false,
        skip_windows: false,
    };

    let config_path = test_dir.join(".testconfig");
    if !exists(&config_path) {
        return Ok(config);
    }

    let content = fs::read_to_string(config_path)?;
    let lines = content.split('\n');

    // Track whether cmd was explicitly set in the config file.
    // In bash, shell is processed first (clears cmd), then cmd is processed
    // (overrides). This means cmd always wins regardless of file order.
    // We match that by deferring the shell→cmd clearing to after the loop.
    let mut cmd_explicitly_set = false;

    for line in lines {
        let trimmed = line.trim();
        if trimmed.is_empty() || trimmed.starts_with('#') {
            continue;
        }

        let Some((key, value)) = trimmed.split_once(':') else {
            continue;
        };

        let key = key.trim();
        let value = value.trim();

        match key {
            "auto_start" => config.auto_start = value != "false",
            "shell" => config.shell = value.to_owned(),
            "cmd" => {
                config.cmd = value.to_owned();
                cmd_explicitly_set = true;
            }
            "stdin" => config.stdin = value.to_owned(),
            "ignore_stderr" => config.ignore_stderr = value != "false",
            "cwd" => config.cwd = value.to_owned(),
            "file_watcher" => {
                if !value.is_empty() {
                    config.file_watcher = value.to_owned();
                }
            }
            // start_args takes everything after "start_args:" with whitespace
            // trimmed. Use .trim() to match bash's awk field-splitting which
            // normalizes leading whitespace. Without this, extra spaces after
            // the colon leave residual whitespace that causes split(/\s+/) to
            // produce empty array elements.
            "start_args" => config.start_args = value.to_owned(),
            "wait_for_recheck" => {
                config.wait_for_recheck =
                    if value == "false" { "false" } else { "true" }.to_owned();
            }
            "skip_saved_state" => config.skip_saved_state = value == "true",
            "saved_state_only" => config.saved_state_only = value == "true",
            "git" => config.git = value == "true",
            "skip_windows" => config.skip_windows = value == "true",
            _ => {}
        }
    }

    if !config.shell.is_empty() && !cmd_explicitly_set {
        config.cmd.clear();
    }

    Ok(config)
}
